text/0000-custom-serialization.md

```diff
@@ -0,0 +1,139 @@
+- Feature Name: custom_serialization
+- Start Date: 2016-10-07
+- RFC PR:
+- Pony Issue:
+
+# Summary
+
+This feature would allow the programmer to specify custom serialization/deserialization that would be run as part of Pony's built-in serialization/deserializaton process. The custom methods would run after the built-in serialization and deserialization, and would allow the programmer to:
+* specify the number of bytes to use for serialization
+* write bytes to the serialization buffer
+* read bytes from the serialization buffer
+
+# Motivation
+
+This is primarily intended as a way to allow programmers to provide systems for serializing and deserializing objects that contain `Pointer` fields. Currently the runtime will raise an error if an object that contains a `Pointer` field is serialized. For example, consider a situation where a Pony program has logic that is implemented in C and has objects that store pointers to data that is used by C code. It is currently impossible to call `Serialise.apply(...)` on the objects to create serialized representations of them. This becomes an especially pressing issue when attempting to write API code for handling user-created objects that may or may not contain `Pointer` fields, depending on the implementation choices made by the user.
+
+# Detailed design
+
+A class providing its own serialization system would need to implement methods for serializing and deserializing data, as well as a method for conveying the number of additional bytes needed for custom serialization of the object. The runtime would call these methods at the appropriate time to generate serialized data and deserialize that data. Pony's built-in serialization would still be performed on the objects, this system is intended to allow *additional* data to be stored in the serialized representation of the object and recovered by deserialization.
+
+## What Gets Serialized And Deserialized
+
+The intent of this system is to allow the programmer to specify a way to use Pony's existing serialization system to work with objects that contain `Pointer` fields to C data structures. Consequently the expectation is that the system would only be used to serialize and deserialize `Pointer` fields, since the other fields are already serialized by Pony's built-in system. However, there is nothing preventing the user from including information from other fields in the serialized representation, nor from using the serialized data to modify non-`Pointer` fields during deserialization.
+
+## Methods
+
+All of the following methods must be implemented for custom serialization:
+* `fun _serialise_space(): USize` -- returns the number of bytes to reserve for custom serialization
```

Comment:
@jemc it depends on what you mean by "next serialized object begins". In the serialized representation, each field contains either a value (if the field stores a numeric value) or a "pointer" to the place in the buffer that represents object that goes in the field. So there isn't really a concept of "next" in the representation, only fields that point to other objects in the buffer. For all objects except `String`s and `Array`s, the size of the object's representation is dictated by the object itself, so there is no need to store the size. In the case of `String`s and `Array`s, the size is stored because these data structures are inherently variable in length.